### ISSUE TITLE
Failing to install sqlc v1.5.0 / latest

### ISSUE THREAD

**Issue by brunoluiz** [kyleconroy/sqlc#654]
## Issue

`go install github.com/kyleconroy/sqlc/cmd/sqlc` seems to be failing due to dolphin changes

## How to reproduce

As there are some C dependencies in sqlc, we have the following target on our `Makefile`:

```
install-generators:
	# more tools...
	tmp_dir=sqlc-$(shell date +'%s') && mkdir -p /tmp/$$tmp_dir && cd /tmp/$$tmp_dir && go mod init tmp && GO11MODULE=on go install github.com/kyleconroy/sqlc/cmd/sqlc && cd - # it has C dependencies that would be a headache using go mod vendor
```

But now it crashes with the following error:

```
go: creating new go.mod: module tmp
go: finding module for package github.com/kyleconroy/sqlc/cmd/sqlc
go: found github.com/kyleconroy/sqlc/cmd/sqlc in github.com/kyleconroy/sqlc v1.5.0
# github.com/kyleconroy/sqlc/internal/engine/dolphin
/Users/brunosilva/go/pkg/mod/github.com/kyleconroy/sqlc@v1.5.0/internal/engine/dolphin/convert.go:197:13: n.Schema undefined (type *"github.com/pingcap/parser/ast".FuncCallExpr has no field or method Schema)
/Users/brunosilva/go/pkg/mod/github.com/kyleconroy/sqlc@v1.5.0/internal/engine/dolphin/convert.go:216:14: n.Offset undefined (type *"github.com/pingcap/parser/ast".FuncCallExpr has no field or method Offset)
```

## Solution for now:

We've changed the script to download a specific version for now

```
install-generators:
	# more tools...
	tmp_dir=sqlc-$(shell date +'%s') && mkdir -p /tmp/$$tmp_dir && cd /tmp/$$tmp_dir && go mod init tmp && go mod edit -require github.com/kyleconroy/sqlc@v1.4.0 && go install github.com/kyleconroy/sqlc/cmd/sqlc && cd - # it has C dependencies that would be a headache using go mod vendor
```

EDIT: Above solution as a bash/shell script (above one is for `Makefile`):

```
tmp_dir=sqlc-$(date +'%s') && \
mkdir -p /tmp/$tmp_dir && \
cd /tmp/$tmp_dir && \
go mod init tmp && \
go mod edit -replace github.com/pingcap/parser=github.com/kyleconroy/parser@v0.0.0-20200728181224-b006e3bf3ed9 && \
go get github.com/kyleconroy/sqlc/cmd/sqlc@v1.5.0 && \
cd - # it has C dependencies that would be a headache using go mod vendor
```

**Comment by kyleconroy** [comment]
Hmm, not sure why that's failing. What version of Go are you using?

I will say that installing via the `go` tool is my least recommended method, as things like this tend to crop up from time to time, especially because sqlc depends on packages that use cgo.

**Comment by brunoluiz** [comment]
My go version:

```
go version go1.14.3 darwin/amd64
```

I would use through brew, but I run generators on our CI (linux) and most people use Macs for development. Hence why we used the `go install` as mentioned.

I had a look today and it seems it is due to this line in `go.mod`:

```
replace github.com/pingcap/parser => github.com/kyleconroy/parser v0.0.0-20200728181224-b006e3bf3ed9
```

If I change add to the `go.mod` on the tmp folder it works fine

**Comment by Ximik** [comment]
Hey. Have the same issue when trying to update to 1.5.0.
```

# github.com/kyleconroy/sqlc/internal/engine/dolphin
../../go/pkg/mod/github.com/kyleconroy/sqlc@v1.5.0/internal/engine/dolphin/convert.go:197:13: n.Schema undefined (type *"github.com/pingcap/parser/ast".FuncCallExpr has no field or method Schema)
../../go/pkg/mod/github.com/kyleconroy/sqlc@v1.5.0/internal/engine/dolphin/convert.go:216:14: n.Offset undefined (type *"github.com/pingcap/parser/ast".FuncCallExpr has no field or method Offset)

```

I've also tried to clean up my GOSRC, which hasn't helped.

go version go1.14.2 linux/amd64
Linux 4.19.44-gentoo

**Comment by brunoluiz** [comment]
@Ximik Are you using just `go get github.com/kyleconroy/sqlc/cmd/sqlc`? If so, I guess you might need to create a temp go module and then add the mod replacement to make it work properly.

The following should work from the terminal:

```
tmp_dir=sqlc-$(date +'%s') && \
mkdir -p /tmp/$tmp_dir && \
cd /tmp/$tmp_dir && \
go mod init tmp && \
go mod edit -replace github.com/pingcap/parser=github.com/kyleconroy/parser@v0.0.0-20200728181224-b006e3bf3ed9 && \
go get github.com/kyleconroy/sqlc/cmd/sqlc@v1.5.0 && \
cd - # it has C dependencies that would be a headache using go mod vendor
```

https://github.com/kyleconroy/sqlc/issues/654#issuecomment-670468435

**Comment by kyleconroy** [comment]
Whoops, I didn't mean to close this when merging #665. 

The problem is that the replace directive only applies inside the sqlc project itself. When you try to bring it in, you need to add the replace directive. Sadly, the replace directive is needed until I get my changes to the MySQL parser merged upstream.

**Comment by warent** [comment]
This is happening for me as well. Looking at the docs:
https://godoc.org/github.com/pingcap/parser/ast#FuncCallExpr


FuncCallExpr is defined as:

```go
type FuncCallExpr struct {
    Tp     FuncCallExprType
    Schema model.CIStr
    // FnName is the function name.
    FnName model.CIStr
    // Args is the function args.
    Args []ExprNode
    // contains filtered or unexported fields
}
```

Trying to access some field `Offset`
https://github.com/kyleconroy/sqlc/blob/master/internal/engine/dolphin/convert.go#L343


But it doesn't seem to be defined:
https://github.com/pingcap/parser/blob/master/ast/functions.go#L351

Nor in the embedded structs:
https://github.com/pingcap/parser/blob/53ac409ed043f89a850043d2440aaaa7dd863a1c/ast/base.go#L92
https://github.com/pingcap/parser/blob/53ac409ed043f89a850043d2440aaaa7dd863a1c/ast/base.go#L64
https://github.com/pingcap/parser/blob/53ac409ed043f89a850043d2440aaaa7dd863a1c/ast/base.go#L20

**Comment by warent** [comment]
It is defined in SelectField though:
https://godoc.org/github.com/pingcap/parser/ast#SelectField

Which is being used elsewhere:
https://github.com/kyleconroy/sqlc/blob/master/internal/engine/dolphin/convert.go#L417

**Comment by kyleconroy** [comment]
Sorry, I should have offered a clearer explanation in my previous comment.

sqlc currently uses my fork (https://github.com/kyleconroy/parser) of pingcap's MySQL parser. The fork adds the `Offset` field to `FuncCallExpr`. This is necessary to support named arguments via `sqlc.arg()`. I'm using the fork via the `replace` directive in go.mod, which apparently doesn't work when installing via `go get`.

I'll be able to move off my fork once https://github.com/pingcap/parser/pull/984 has been merged. Until then, builds using `go get` will fail. Please install via Docker or the download links in the README in the time being.

**Comment by warent** [comment]
Oh I see! Thanks for the explanation & quick response. I just figured it out after cloning the repo and successfully building from source, leading me to find the go.mod replacement 😄

**Comment by rochacon** [comment]
I use `sqlc` in different projects with  different versions. Calling it via `//go:generate go run .../cmd/sqlc generate` is a way to avoid having the binary in a different version then the one in the project's go.mod.
Including the same [`replace` directive from the v1.5.0 release](https://github.com/kyleconroy/sqlc/blob/v1.5.0/go.mod#L20) in my `go.mod` file was the workaround I took.

**Comment by kyleconroy** [comment]
Re-opening this until the next version is released.

**Comment by kyleconroy** [comment]
This has been fixed since the release of v1.6.0
